plugin-server/src/worker/ingestion/event-pipeline/cookielessServerHashStep.ts

```diff
@@ -0,0 +1,303 @@
+        sessionInfo = { s: sessionId, t: timestampMs }
+        await runner.hub.db.redisSet(sessionRedisKey, sessionInfo, 'cookielessServerHashStep', SESSION_TTL_SECONDS)
+    } else {
+        // otherwise, update the timestamp
+        await runner.hub.db.redisSet(
+            sessionRedisKey,
+            { s: sessionInfo.s, t: timestampMs },
+            'cookielessServerHashStep',
+            SESSION_TTL_SECONDS
+        )
+    }
+
+    event.properties['$session_id'] = sessionInfo.s
+
+    return [event]
+}
+
+function getProperties(
+    event: PluginEvent,
+    timestamp: string
+): {
+    userAgent: string | undefined
+    ip: string | undefined
+    host: string | undefined
+    timezone: string
+    timestampMs: number
+    teamId: number
+} {
+    const userAgent = event.properties?.['$raw_user_agent']
+    const ip = event.properties?.['$ip']
+    const host = event.properties?.['$host']
+    const timezone = event.properties?.['$timezone'] || TIMEZONE_FALLBACK
+    const timestampMs = DateTime.fromISO(timestamp).toMillis()
+    const teamId = event.team_id
+
+    return { userAgent, ip, host, timezone, timestampMs, teamId }
+}
+
+const localSaltMap: Record<string, Uint32Array> = {}
+const mutex = new ConcurrencyController(1)
+
+export async function getSaltForDay(db: DB, timestamp: number, timeZone: string): Promise<Uint32Array> {
+    // get the day based on the timezone
+    const parts = new Intl.DateTimeFormat('en', {
+        timeZone,
+        year: 'numeric',
+        month: '2-digit',
+        day: '2-digit',
+    }).formatToParts(new Date(timestamp))
+    const year = parts.find((part) => part.type === 'year')?.value
+    const month = parts.find((part) => part.type === 'month')?.value
+    const day = parts.find((part) => part.type === 'day')?.value
+    const yyyymmdd = `${year}-${month}-${day}`
+
+    // see if we have it locally
+    if (localSaltMap[yyyymmdd]) {
+        return localSaltMap[yyyymmdd]
+    }
+
+    // get the salt for the day from redis, but only do this once for this node process
+    return mutex.run({
+        fn: async (): Promise<Uint32Array> => {
+            // check if we got the salt while waiting for the mutex
+            if (localSaltMap[yyyymmdd]) {
+                return localSaltMap[yyyymmdd]
+            }
+
+            // try to get it from redis instead
+            const saltBase64 = await db.redisGet<string | null>(
+                `cookieless_salt:${yyyymmdd}`,
+                null,
+                'cookielessServerHashStep'
+            )
+            if (saltBase64) {
+                const salt = base64StringToUint32Array(saltBase64)
+                localSaltMap[yyyymmdd] = salt
+                return salt
+            }
+
+            // looks like we'll need to create a new one, but check the date is within range
+            if (!isCalendarDateValid(yyyymmdd)) {
+                throw new Error('Date is out of range')
+            }
```

Comment:
This could probably be outside the mutex call, right? We can detect it's invalid from the main function

Reply:
Done